### ISSUE TITLE
Weird splitting when generating sierpinski triangle with <10 iterations

### ISSUE THREAD

**Issue by Robin-Dillen** [Robin-Dillen/3DEngine#6]
![l_systems021](https://user-images.githubusercontent.com/72447513/111028686-79970300-83f8-11eb-9278-de01bbb6047d.png)
 On the left side, gets worse the more iteration you have. Probably some rounding error..

**Comment by JamesBremner** [comment]
What code did you use to produce this?

**Comment by Robin-Dillen** [comment]
The current code in the main branch normally

**Comment by JamesBremner** [comment]
That sounds like you mean the tip of the master branch.  Currently that is 25fceab9baa3576bafeff2e6f5f27ba6d9b799a8 which was committed a minute ago.  This issue was opened yesterday, so that cannot be the case.

**Comment by Robin-Dillen** [comment]
I couldn't really fix it but I made some other changes. The problem still occurs though.

**Comment by JamesBremner** [comment]
I explained before that you should not hack away at the master branch.  Until you are sure that you have not broken anything development commits should go into their own branch.  Only once everything is working should the master branch be updated from the development.

( I think that you have said you are working on something for a course in coding.  Is that correct?  Where is this course being taught.  Are they teaching you anything about collaborative coding? )

**Comment by JamesBremner** [comment]
> I couldn't really fix it

So you are working on this issue, then.

1. Assign the issue to yourself and labell it accepted, so that others know you are working on it.
2. Stop hacking away at the master branch.  It make it impossible for anyone else to collaborate with you.

**Comment by Robin-Dillen** [comment]
I follow a bachelors degree at the University of Antwerp(first year), they haven't said a thing about github or collaborative coding for that matter.

**Comment by JamesBremner** [comment]
In the real world, coding is a team sport.  You need to know how to play!

**Comment by JamesBremner** [comment]
Your course instructor should be encouraging the students to get together in small groups and do assignments as a team.  Ask why this is not happening.  Antwerp sounds like a serious place, but they seem to be horribly out of date.

**Comment by Robin-Dillen** [comment]
I know, but i guess we gotta figure it out ourselfs, but since the hacking away works enough, I just keep doing it

**Comment by JamesBremner** [comment]
You are learning the wrong stuff, if you want to get a job in software engineering.  No one will pay you to hack at their software.

**Comment by Robin-Dillen** [comment]
https://www.youtube.com/watch?v=3RjQznt-8kE&list=PL4cUxeGkcC9goXbgTDQ0n_4TBzOO0ocPR&ab_channel=TheNetNinja
Would this be a good tutorial to follow?
And would it be good to maybe clear the repo and start over, so that all the hacks are gone?

**Comment by JamesBremner** [comment]
> And would it be good to maybe clear the repo and start over, so that all the hacks are gone?

No need.  That would be to discard all the work you have already invested.  The thing is to improve your work habits as you go along.  Software engineering is always about learning from mistakes and then not repeating them.

**Comment by Robin-Dillen** [comment]
> Your course instructor should be encouraging the students to get together in small groups and do assignments as a team. Ask why this is not happening. Antwerp sounds like a serious place, but they seem to be horribly out of date.

Maybe to will teach us later on but I doubt it.

**Comment by Robin-Dillen** [comment]
So concreetly, I should only push **working** code to the main branch. If I have bugs that need to be fixed I should work on them in another branch?

**Comment by JamesBremner** [comment]
You should always work in a development branch, whatever you are doing.

When you are satisfied with your work, that is when all tests pass, you should update the master branch from the development branch 

Work should never be directly pushed to the master branch.  If you do that, you wreck whatever work other people might be doing.

**Comment by JamesBremner** [comment]
Here is an example of development tree on a project I recently collaborated on

![image](https://user-images.githubusercontent.com/2046227/111074874-8f213100-84bb-11eb-9074-f96e4a351304.png)

**Comment by JamesBremner** [comment]
Some things to notice.

* Development work is done in separate branches
* Occasionally a quick fix is pushed to the master ( trunk here ) but this should only be done very occasionally when the fix is timy. ( in this example I left some debugging instrumentation enabled so I did a one line xhange to disable it.
* Commits reference the issue where the work is discussed.

**Comment by Robin-Dillen** [comment]
I think I get it now. So when I'm going to try the fix the splitting issue. Or when I try to make the code faster(in general edit or add a feature). I create a branch. Then when I'm happy with the feature (it works as you said) I push it(is it pushing?). When I'm working in a team other can review the branch and in the end they can accept it.

**Comment by Robin-Dillen** [comment]
So for example, I noticed my code crashes when I try to make a sierpinski triangle with 3 iterations. So I make a branch let's say (fix-3-iterations), and then I work on it until it's fixed, and then I push it and merge it with the main branch.
When I create the issue I mention this new branch (and close it since I'm working on it?)

**Comment by JamesBremner** [comment]
> Then when I'm happy with the feature (it works as you said) I push it(is it pushing?).

You push code ( from your development machine ) to a development branch ( in the repository )

You merge a development branch ( in the repo ) into the master branch ( also in the repo )

Merging is an important skill which you should practice.  

- Create a development branch
- Check out the branch
- Make a few small commits to the branch
- Push commits to the branch
- Merge branch into master

Try this a few times untill you are comfortable and confident doing this.  Then you can boast about it at your first job interview.

**Comment by JamesBremner** [comment]
>  then I work on it until it's fixed, and then I push it and merge it with the main branch.

That would work for a small bug.  For a complicated bug or a new feature it goes more like this:

- Work
- Push
- Work
- Pust
- Test
- Work
- Push
- Test
- Merge

**Comment by JamesBremner** [comment]
I notice that you still have not assigned and labelled this issue!

**Comment by Robin-Dillen** [comment]
> * Merge branch into master
I'm working in clion. Do I do this in clion, online or the console

**Comment by JamesBremner** [comment]
>When I create the issue I mention this new branch (and close it since I'm working on it?)

Creating the issue comes first!

Then you can refence the issue in the branch name and in the commits.

**Comment by Robin-Dillen** [comment]
But isn't this the issue or do I need to redo it?

**Comment by JamesBremner** [comment]
> I'm working in clion. Do I do this ( you mean merge into master ) : in clion, online or the console

Merging is done in the repo client - that is github desktop.

**Comment by JamesBremner** [comment]
> But isn't this the issue or do I need to redo it?

This issue is about "wierd splitting".  The crash sounds like it should be a new issue.

**Comment by Robin-Dillen** [comment]
> This issue is about "wierd splitting". The crash sounds like it should be a new issue.

Yes it is, it's still a little confusing sometimes.

**Comment by Robin-Dillen** [comment]
I hope I did everything(correct)

**Comment by JamesBremner** [comment]
> I hope I did everything(correct)

What sis you do?

**Comment by Robin-Dillen** [comment]
I assigned myself to the issues and added a label. Then I already fixed an issue(and hopefuly did everything correct with the branches and so on)

**Comment by JamesBremner** [comment]
Looks good

![image](https://user-images.githubusercontent.com/2046227/111078398-90f2f080-84cb-11eb-8397-edc22169ec4a.png)

**Comment by Robin-Dillen** [comment]
The issue is due to rounding errors, and I do really need to to be solved, in the future i could reopen the issue if i want to
